Feature/shard path multiplexing

undertome opened the pull request:
This PR introduces changes that allow a user to specify multiple paths for storing historical shards. The changes are based in the `DatabaseShardImp` class and in the shards unit tests. Please read [the documentation](https://github.com/undertome/rippled/blob/feature/shard-path-multiplexing/src/ripple/nodestore/ShardStoragePaths.md) to get an idea of the changes involved.

MarkusTeufelberger commented on src/ripple/nodestore/ShardStoragePaths.md:
```diff
@@ -0,0 +1,56 @@
+
+# List of paths for storing older shards.
+[historical_shard_paths]
+/dev/sdb1
```
I doubt that shards will be stored on raw devices? Might need a better path like `/mnt/disk1` for example.

undertome replied:
Done

MarkusTeufelberger commented on src/ripple/nodestore/ShardStoragePaths.md:
```diff
@@ -0,0 +1,56 @@
+
+### Selecting a Historical Storage Path
+
+When storing historical shards, if multiple historical paths are provided, the paths will be selected in a round-robin fashion. By using all available storage devices, we increase our chances for concurrent access to stored shards. Naturally, a storage path is included in the round-robin distribution only if it has enough storage capacity to hold the next shard.
```
This would cause issues if I for example have 3 paths configured, the respective disks start getting full and I add a 4th one. I would recommend to store the next shard at the disk that has the least space used by shards (still an issue if disk sizes are different, but less problematic than the "proper" solution of choosing the disk that has the most free space) - or, if getting sizes is too difficult, then at least choosing the disk that has the least amount of shard files stored, with a random selection (or just the order in the config file) as tie-breaker.

undertome replied:
The selection process outlined here is actually outdated - during the implementation I decided to go with random selection. The doc in the most recent push has been updated with that change.

I think the random selection is sufficient for our needs. It's simple (which is good) and does the job well enough. If the motivation for using capacity-based selection is to increase performance by reducing file fragmentation, I don't think its worth implementing. For any server that stores limited history, shard fragmentation is inevitable and eventually all of its historical shards will be fragmented. And for a server that stores full history, in theory shard fragmentation won't occur. In either case I don't see a reason for space-based selection.

seelabs commented on src/ripple/nodestore/ShardStoragePaths.md:
```diff
@@ -0,0 +1,73 @@
+
+## Overview
+
+The shard database stores validated ledgers in logical groups called shards. As of June 2020, a shard stores 16384 ledgers by default. In order to allow users to store shards on multiple devices, the shard database can be configured with several file system paths. Each path provided should refer to a directory on a distinct filesystem, and no two paths should ever correspond to the same filesyste ...
```
Please limit lines to 80 characters. Rational: one advantage of markdown is it may be read in raw form. The long lines make it more difficult to read in the terminal and in some editors. Also, long lines make it more difficult to leave review comments localized to the area of the issue.

seelabs replied:
For the sentence:
```
Each path provided should refer to a directory on a distinct filesystem, 
and no two paths should ever correspond to the same filesystem.
```
We should be able to check this condition in code - for linux at least - and either error out or warn if the condition isn't met.

undertome replied:
> Please limit lines to 80 characters. Rational: one advantage of markdown is it may be read in raw form. The long lines make it more difficult to read in the terminal and in some editors. Also, long lines make it more difficult to leave review comments localized to the area of the issue.

Done

undertome replied:
> For the sentence:
> 
> ```
> Each path provided should refer to a directory on a distinct filesystem, 
> and no two paths should ever correspond to the same filesystem.
> ```
> 
> We should be able to check this condition in code - for linux at least - and either error out or warn if the condition isn't met.

Added this check for Linux. Errors out during initialization.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+    }
+}
+
+boost::optional<bool>
```
`optional<bool>` is error prone to use. It's too easy to use the outer `bool` instead of the inner `bool`. I'd recommend replacing the inner `bool` with an enum.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+bool
+DatabaseShardImp::sufficientStorage(
+    std::uint32_t numShards,
+    bool isHistorical,
```
Replacing the `bool` with an enum would make the calling sites clearer.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+                           std::uint64_t(numShards));
+
+                       return numShards == 0;
+                   }) != capacities.end();
```
While clever, we aren't using `find_if` to find anything. It's likely people will need to decode this. What do you think about calculating the available number of shards and checking if there's enough space that way instead. I.e.
```

        std::uint64_t availSpacsShards = 0;
        for (auto const c : capacities)
        {
            availSpacsShards += c / avgShardFileSz_;
        }
        return availSpacsShards >= numShards;
```
That code does not exit early (tho it's easy enough to do that). But given the size of `capacities` it won't matter.

undertome replied:
Changed to a for loop.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+                [&capacities](boost::filesystem::path const& path) {
+                    capacities.push_back(
+                        boost::filesystem::space(path).available);
+                });
```
A range for loop feels simpler for this. I.e.
```
            capacities.reserve(historicalPaths_.size());
            for (auto const& p : historicalPaths_)
                capacities.push_back(boost::filesystem::space(p).available);
```
Note, even if we end up keeping the `for_each`, we should still `reserve`

undertome replied:
Added reserve.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
-        return boost::filesystem::space(dir_).available;
+        std::vector<std::uint64_t> capacities;
+
+        // Get the available storage for each historical path
```
Nit: We usually put comment pertaining to the block inside the block, not before the block (for the `else` block as well)

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,12 +1484,48 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+                           std::uint64_t(numShards));
+
+                       return numShards == 0;
+                   }) != capacities.end();
```
Since you're modifying this file, Line 1534 (not part of this changeset) should be return false not return 0.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -292,10 +309,15 @@ DatabaseShardImp::prepareShard(std::uint32_t shardIndex)
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    bool isHistoricalShard = shardIndex < getShardBoundaryIndex(lock);
+    auto numHistoricalShards = countHistoricalShards(lock);
```
The above two vars could be `const`

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -392,15 +414,28 @@ DatabaseShardImp::importShard(
-    auto shard{std::make_unique<Shard>(app_, *this, shardIndex, j_)};
+    auto shard{std::make_unique<Shard>(app_, *this, shardIndex, j_, dstDir)};
+
+    dstDir /= std::to_string(shardIndex);
```
We convert `shardIndex` to a string in two places. We may want to create a local var for this (we do it here and line 431).

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1110,74 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
     using namespace boost::filesystem;
-    if (!get_if_exists<path>(section, "path", dir_))
-        return fail("'path' missing");
+    if (!get_if_exists<path>(section, "main_storage_path", dir_))
```
Does this break existing config files?

undertome replied:
Yes

undertome replied:
Rolled back.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1110,74 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
-        maxFileSz_ = sz << 30;
+        std::string hPathError = "";
+
+        if (std::find_if(
```
I'd rewrite this so it doesn't use `find_if`. Rational: 1) we aren't finding anything, so it's easy to misread the code, 2) Normally the lambdas to `find_if` are pure, here the lambda is creating directories 3) Even if the lambda were pure, an explicit range for loop is something most developers are used to seeing for code like this. Unless there's a compelling reason, it's good to use idioms other developers expect.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1298,12 +1430,14 @@ DatabaseShardImp::setFileStats()
     avgShardFileSz_ = (numShards == 0 ? fileSz_ : fileSz_ / numShards);
 
-    if (fileSz_ >= maxFileSz_)
+    if (auto count = countHistoricalShards(lock); count >= maxHistoricalShards_)
```
Nit: `const`

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+    return std::count_if(
+        shards_.begin(),
+        shards_.end(),
+        [recentShardBoundaryIndex](decltype(shards_)::value_type const& entry) {
```
Could use `auto const& entry` for the param type here and get rid of the `decltype`

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+
+        //////////////////////////////////////////////////////////////////
+        // Lambdas used inside of this method:
+        auto const removeShard = [this](std::uint32_t const shardIndex) {
```
Consider adding return types for all these lambdas. We often omit return types, for when there are several helper functions like this I'd prefer to add them.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+        [this, &potentialPaths](boost::filesystem::path const& path) {
+            if (boost::filesystem::space(path).available >= avgShardFileSz_)
+                potentialPaths.push_back(path);
+        });
```
Consider using a range based for loop instead. It's shorter, and more common than `for_each` in our code.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1595,266 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+    if (historicalPaths_.empty())
+        return dir_;
+
+    std::array<boost::filesystem::path, 1> historicalShardPath;
```
Instead of a single element array, consider just `boost::filesystem::path historicalShardPath;` When you need an iterator for `std::sample` take the address. I.e.
```
    std::sample(
        potentialPaths.begin(),
        potentialPaths.end(),
        &historicalShardPath,
        1,
        default_prng());
```

undertome replied:
Done

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DeterministicShard.h:
```diff
@@ -61,7 +61,8 @@ class DeterministicShard
         uint256 const& lastHash,
-        beast::Journal j);
+        beast::Journal j,
+        boost::filesystem::path const& dir);
```
`journal` should be the last parameter

undertome replied:
I think this will get removed since `DeterministicShards` are getting pulled from the release.

undertome replied:
Removed.

seelabs commented on src/ripple/nodestore/impl/Shard.h:
```diff
@@ -56,13 +56,17 @@ class Shard final
         std::uint32_t index,
-        beast::Journal j);
+        beast::Journal j,
+        boost::filesystem::path const& dir = "");
```
`journal` are usually last parameter. In this, default parameters won't work with that. We'll either two functions or we can keep `journal` where it is. Your call.

undertome replied:
Added a second constructor

seelabs commented on src/ripple/rpc/impl/ShardArchiveHandler.cpp:
```diff
@@ -37,9 +37,9 @@ boost::filesystem::path
 {
     return get(config.section(ConfigSection::shardDatabase()),
-               "download_path",
+               "partial_shard_download_path",
```
Does this break existing config files?

undertome replied:
I guess that depends on what you mean by break. A server whose config file only provides the oudated property name (`download_path`) will use a default path as the `partial_download_path` which may or may not reflect whatever path was specified with the outdated name.

undertome replied:
Rolled back.

undertome commented:
@mDuo13 This PR contains changes to two `shard_db` properties:

1.  `path` -> `main_storage_path` **(breaking change)**
2. `download_path` -> `partial_shard_download_path`

miguelportilla commented on src/test/nodestore/DatabaseShard_test.cpp:
```diff
@@ -162,6 +160,7 @@ randInt(Engine& engine, Integral max)
 class DatabaseShard_test : public TestBase
 {
     static constexpr std::uint32_t maxSizeGb = 10;
+    static constexpr std::uint32_t maxHistoricalShards = 100;
```
This file needs to `#include <numeric>`

miguelportilla commented on src/test/nodestore/DatabaseShard_test.cpp:
```diff
@@ -1185,6 +1191,320 @@ class DatabaseShard_test : public TestBase
+            // should be stored at the main path
+            BEAST_EXPECT(mainPathCount == 2);
+
+            auto const historicalPathCount = std::accumulate(
```
This file needs to `#include <numeric>`

ximinez replied:
The Windows build is broken without it. I put this branch on top of the PR that fixes the Travis builds, and you can see all the Windows builds fail. https://travis-ci.com/github/ximinez/rippled/builds/177519102

undertome replied:
Oh, good catch. Done.

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1627,273 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+                            shard->getDir().string(),
+                            dst / std::to_string(shardIndex));
+                    }
+                    catch (std::exception const& e)
+                    {
+                        JLOG(j_.error())
+                            << "shard " << shardIndex
+                            << " failed to move to historical storage";
+
+                        return;
+                    }
```
`e` is unused. Either remove it, or output `e.what()` as part of the error message.

undertome replied:
Done

undertome commented on src/ripple/nodestore/ShardStoragePaths.md:
```diff
@@ -0,0 +1,123 @@
+chances for concurrent access to stored shards, however as of June 2020
+concurrent shard access is not implemented. Lastly, a storage path is included
+in the random distribution only if it has enough storage capacity to hold the
+next shard.
+## Shard Acquisition
```
Done

undertome commented on src/ripple/net/ShardDownloader.md:
```diff
@@ -175,9 +175,9 @@ The `download_path` field of the `shard_db` entry is used to determine where to
-path=/var/lib/rippled/db/shards/nudb
-download_path=/var/lib/rippled/db/shards/
-max_size_gb=50
+main_storage_path=/var/lib/rippled/db/shards/nudb
+partial_shard_download_path=/var/lib/rippled/db/shards/
+max_historical_shards=50
```
> I don't like some of these name changes:
> 
>     1. `path` to `main_storage_path` feels redundant. A path is clearly talking about storage, and if it doesn't have a prefix, it makes sense to assume that it's the "main" path. Additionally, some operators were already using shards, and now they have to make an arbitrary change that doesn't really add any value.
> 
>     2. `download_path` to `partial_shard_download_path` isn't as eggregious, but we're already in the `[shard_db]` section, so "shard" in the name is also redundant. "partial" also feels unnecessary, though I suppose you could make an argument that it's more descriptive since fully downloaded shards aren't left there, but that feels like a really unnecessary addition since we don't have a competing `complete_download_path`.
> 
> 
> For both `path` and `download_path`, since existing configs use the old values, and these name changes don't seem to add anything that can't be added with updated and clarified docs, I think these name changes should be rolled back
> 

I don't like renaming the two path keys either. It was suggested by @scottschurr on #3927 and I disagreed with the suggestion. See [this comment thread](https://github.com/ripple/rippled/pull/3297#discussion_r402649904). I'm fine with reverting the path names but I think we should arrive at a consensus to avoid further thrashing. Pinging @miguelportilla

undertome replied:
> I'm OK with switching from a size to a count, but I'm curious what the motivation for the change was. The reason it might be a problem is that it's pretty easy for an operator to look at their device like, "I have 100Gb free on this disk. I'll let shards have 60Gb of it." Whereas there's no real way to definitively say how much space 50 shards will take up, especially since they vary by size. Would it be worth adding an optional "minimum free space" field so that any download stops if the disk gets less than that amount of space?

The decision to control storage with a shard count rather than a byte count came from @miguelportilla. This was partly motivated by a desire to mimic the node store's approach of specifying ledger counts to limit db size, placing the burden of ensuring the availability of sufficient storage capacity on the user. Also note that the `DatabaseShardImp` does provide some help by issuing a warning if the available capacity is insufficient to store the full quantity of shards specified by the shard count. However, this warning is just meant to be supplementary and we expect the user to be responsible for providing ample storage for any server storing shard history.

Currently, the `DatabaseShardImp` already implements some notion of "minimum free space" by estimating the average size of a shard and minimally requiring that this much space must be available before storing an additional shard. Unfortunately this messy calculation is not ideal as shard sizes vary and are positively correlated with the shard index (later shards are typically larger in size). We have tentative plans to improve this somewhat by adding a sizeable buffer to this minimum size requirement (this change is partly inspired by the addition of `DeterministicShards`, as the capacity requirements during the finalization of shards may be double the size of the finalized shard). So in short, we're in the process of improving capacity management, but as of yet we have no plans of doing this with configuration properties.

undertome replied:
> I don't like some of these name changes:
> 
>     1. `path` to `main_storage_path` feels redundant. A path is clearly talking about storage, and if it doesn't have a prefix, it makes sense to assume that it's the "main" path. Additionally, some operators were already using shards, and now they have to make an arbitrary change that doesn't really add any value.
> 
>     2. `download_path` to `partial_shard_download_path` isn't as eggregious, but we're already in the `[shard_db]` section, so "shard" in the name is also redundant. "partial" also feels unnecessary, though I suppose you could make an argument that it's more descriptive since fully downloaded shards aren't left there, but that feels like a really unnecessary addition since we don't have a competing `complete_download_path`.
> 
> 
> For both `path` and `download_path`, since existing configs use the old values, and these name changes don't seem to add anything that can't be added with updated and clarified docs, I think these name changes should be rolled back
> 

Rolling back.

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1122,89 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
-        std::uint64_t sz;
-        if (!get_if_exists<std::uint64_t>(section, "max_size_gb", sz))
-            return fail("'max_size_gb' missing");
+        std::uint64_t count = 0;
+        get_if_exists<std::uint64_t>(section, "max_historical_shards", count);
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1122,89 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
+            if (dir_ == dir)
+            {
+                return fail(
+                    "a 'historical_shard_path' value == "
+                    "'main_storage_path'");
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1122,89 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
+                    "failed to create 'historical_shard_path': " +
+                    dir.string());
+            }
+
```
Moved creation to init

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1070,23 +1122,89 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
+        std::sort(values.begin(), values.end());
+        values.erase(std::unique(values.begin(), values.end()), values.end());
+
+        std::unordered_set<int> filesystemIDs;
```
I thought of that but I didn't want to include another preprocessor directive just for that.

undertome replied:
Done

ximinez replied:
I suppose it could go either way. I didn't notice a compiler warning about an unused variable, so maybe it optimized it away.

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1245,6 +1363,47 @@ DatabaseShardImp::finalizeShard(
+
+            if (isHistoricalShard)
+            {
+                if (!historicalPaths_.empty())
+                {
+                    // This is a historical shard that wasn't
+                    // placed at a separate historical path
+                    if (shard->getDir().parent_path() == dir_)
+                    {
+                        JLOG(j_.warn())
+                            << "shard " << shardIndex
+                            << " is not stored at a historical path";
+                    }
+                }
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1245,6 +1363,47 @@ DatabaseShardImp::finalizeShard(
+                            << " is not stored at a historical path";
+                    }
+                }
+            }
+
+            // Not a historical shard
+            else
+            {
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1298,12 +1457,16 @@ DatabaseShardImp::setFileStats()
     }
-    else if (maxFileSz_ - fileSz_ > available())
+    else if (
+        maxHistoricalShards_ > count &&
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,19 +1513,58 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+            // number of shards. For this to work properly,
+            // each historical path must correspond to a separate
+            // physical device or filesystem.
+            numShards -=
+                std::min(capacity / avgShardFileSz_, std::uint64_t(numShards));
+
+            if (numShards == 0)
+                return true;
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1627,273 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+
+            if (auto it = shards_.find(shardIndex); it != shards_.end())
+            {
+                if (auto& shard = it->second.shard; shard)
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1627,273 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+
+                    try
+                    {
+                        // Move the shard directory to the new path
+                        boost::filesystem::rename(
+                            shard->getDir().string(),
+                            dst / std::to_string(shardIndex));
```
This concern relates to a problem the `DatabaseShardImp` had wherein validating completed shards could cause the server to seize up. This was fixed by creating a background thread that would handle shard finalization. This same background thread also handles shard migration, so it shouldn't interfere with consensus call paths.

ximinez replied:
That's exactly what I was hoping to hear. :grin:

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -222,8 +227,11 @@ class DatabaseShardImp : public DatabaseShard
-    // Maximum storage space the shard store can utilize (in bytes)
-    std::uint64_t maxFileSz_;
+    // Maximum number of historical shards to store.
+    std::uint64_t maxHistoricalShards_;
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -286,9 +307,16 @@ class DatabaseShardImp : public DatabaseShard
-    available() const;
+    // Returns true if the filesystem has enough storage
+    // available to hold the specified number of shards.
+    // If isHistorical is set, that means that the shard(s)
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1627,273 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+{
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    auto const boundaryIndex = getShardBoundaryIndex(lock);
+    auto const isHistoricalShard = shardIndex < boundaryIndex;
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -292,10 +314,19 @@ DatabaseShardImp::prepareShard(std::uint32_t shardIndex)
-    if (avgShardFileSz_ > available())
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    bool const isHistoricalShard = shardIndex < getShardBoundaryIndex(lock);
```
Since it's only a one liner in just two spots, I'm inclined to leave as is.

ximinez commented:
@undertome In the future, please do not rewrite history and force push a single commit over and over. Instead, add additional commits onto your branch, starting the commit message with "[FOLD]" to indicate that these commits will need to be squashed once the PR is passed or merged. This makes my job as reviewer exponentially more manageable for several reasons, including I can confirm that you addressed all issues, I can see precisely what changed since the last time I looked at the PR, and I can focus on those changes and lower the risk of missing something that gets "buried" in the original change set. Thanks!

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -82,95 +86,128 @@ DatabaseShardImp::init()
+                    {
+                        if (!is_directory(path))
+                        {
+                            JLOG(j_.error()) << "each path must be a "
+                                                "directory";
```
How about
```
JLOG(j_.error()) << path << " must be a directory";
```
?

undertome replied:
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1245,6 +1355,43 @@ DatabaseShardImp::finalizeShard(
+                // Not a historical shard. Shift recent shards
+                // if necessary
+                checkRecentShards(lock);
+                assert(shardIndex - boundaryIndex <= 1 || !boundaryIndex);
```
I know it's just an assert, but when using short-circuit operators, I usually like to do the "easier" check first. So `assert(!boundaryIndex || shardIndex - boundaryIndex <= 1);`

undertome replied:
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,19 +1501,59 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+
+        for (std::uint64_t const capacity : capacities)
+        {
+            // Leverage all the historical shard paths to
+            // see if collectively they can fit the specified
+            // number of shards. For this to work properly,
+            // each historical path must correspond to a separate
+            // physical device or filesystem.
```
It's not a perfect check, but how hard would it be to check for duplicates when you're populating `capacities`? If two different paths have the same number of bytes available, there's a chance that they're the same device. It's not guaranteed, so don't fail or anything, but you could log a warning message. Something like `JLOG(j_.warn()) << path1 << " and " << path2 << " both have " << capacity << " bytes free, and may be the same device. Each configured historical storage path should be on a unique device.`

You could even make it `#if ! BOOST_OS_LINUX`

undertome replied:
Done

undertome commented:
> @undertome In the future, please do not rewrite history and force push a single commit over and over. Instead, add additional commits onto your branch, starting the commit message with "[FOLD]" to indicate that these commits will need to be squashed once the PR is passed or merged. This makes my job as reviewer exponentially more manageable for several reasons, including I can confirm that you addressed all issues, I can see precisely what changed since the last time I looked at the PR, and I can focus on those changes and lower the risk of missing something that gets "buried" in the original change set. Thanks!

Sure thing!

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,19 +1500,106 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+#endif
+
+            // Get the available storage for each historical path
+            std::for_each(
```
What is your criteria for using `std::for_each` vs. a range based for loop? Most of rippled uses range based for loops (and I prefer them as well). Unless there's a non-stylistic reason for this, I'd prefer we stick to range based for loops. (To be clear, there's nothing wrong with `std::for_each`, it just seems like there's a better way to do that now that we have range based for loops).

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1410,6 +1647,12 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+
+        if (shard->index() == secondLatestShardIndex_)
+            secondLatestShardIndex_ = boost::none;
+
```
Not in this changeset, but it looks like the `shard` parameter should be passed by `const&` and not value (unless I'm missing the reason).

undertome replied:
It should be a reference.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1662,271 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
     setFileStats();
```
Also not in this changeset, but what is the point of the `reset` on the previous line? If there is one it should be documented, if there isn't it should be removed.

undertome replied:
This causes the shard to be removed from the `DatabaseShardImp`. This is why it's important for the shard to be passed by reference.

undertome replied:
Added a comment.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1662,271 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+        auto const separateHistoricalPath = !historicalPaths_.empty();
+
+        //////////////////////////////////////////////////////////////////
+        // Lambdas used inside of this method:
```
It's a bit of a personal preference, but I prefer not to include comments that are obvious from the code. I.e there's no need to point out that these are lambdas used in this method.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1419,6 +1662,271 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+        potentialPaths.end(),
+        &historicalShardPath,
+        1,
+        default_prng());
```
Instead of a random sample, why not choose the one with the most space available?

undertome replied:
It's not necessary - see [this comment thread](https://github.com/ripple/rippled/pull/3463#discussion_r447239944).

undertome commented on src/ripple/nodestore/ShardStoragePaths.md:
```diff
@@ -0,0 +1,124 @@
+several file system paths. Each path provided should refer to a directory on a
+distinct filesystem, and no two paths should ever correspond to the same
+filesystem. Violating this restriction will cause the server to inaccurately
+estimate the amount of space available for storing shards.
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -242,6 +250,19 @@ class DatabaseShardImp : public DatabaseShard
     // File name used to mark shards being imported from node store
     static constexpr auto importMarker_ = "import";
 
+    //////////////////////////////////////////////////////////////////
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+    // Returns the index that represents the logical
+    // partition between historical and recent shards
+    std::uint32_t
+    getShardBoundaryIndex(std::lock_guard<std::mutex> const&);
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+    getShardBoundaryIndex(std::lock_guard<std::mutex> const&);
+
+    std::uint32_t
+    countHistoricalShards(std::lock_guard<std::mutex> const&);
```
Base on this and another renaming suggestion you made, it seems that the convention is to came the method after the variable returned, rather than the action its performing - is that the case?

undertome replied:
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+    // shards as new shards become available on the network.
+    // Older shards are moved to a historical shard path.
+    void
+    checkRecentShards(std::lock_guard<std::mutex> const& lock);
```
I think that's misleading as well since the majority of calls to this function will not result in any relocation.

undertome replied:
Changed to `relocateOutdatedShards`

seelabs replied:
Yep, that's a better name.

ximinez replied:
The new `relocateOutdatedShards` name is better, since it makes it clear _which_ shards will be relocated, but just FYI another convention I've seen is "maybe", so `maybeRelocateShards`.

undertome replied:
That's funny I actually considered "maybe" but wasn't sure if that would be accepted.

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+
+    // Checks whether the shard can be stored. If
+    // the new shard can't be stored, returns
+    // boost::none. Otherwise returns a bool
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+    // placed in a separate directory for historical
+    // shards.
+    boost::optional<PathDesignation>
+    expectNewShard(
```
I agree that's a better name.

undertome replied:
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+        std::lock_guard<std::mutex> const& lock);
+
+    boost::filesystem::path
+    chooseHistoricalPath();
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -297,6 +326,38 @@ class DatabaseShardImp : public DatabaseShard
+    chooseHistoricalPath();
+
+    boost::filesystem::path
+    chooseHistoricalPath(std::lock_guard<std::mutex> const&);
```
Done

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -82,95 +86,127 @@ DatabaseShardImp::init()
+                catch (...)
+                {
+                    JLOG(j_.error())
+                        << "failed to create path: " + path.string();
```
The try loops are a preexisting convention. Seems like a case of six of one... so I'm inclined to leave them as they are.

seelabs replied:
I'm OK leaving them.

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -82,95 +86,127 @@ DatabaseShardImp::init()
                 {
-                    continue;
-                }
+                    if (!is_directory(d))
```
This is already protected by a `try` and `catch` block. However, I'm wondering if the catch statement should return false. I didn't write that part so I will have to look into it.

undertome replied:
I added a `return false` statement to the catch statement

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1074,19 +1143,59 @@ DatabaseShardImp::initConfig(std::lock_guard<std::mutex>&)
 
-        // Convert to bytes
-        maxFileSz_ = sz << 30;
+#if BOOST_OS_LINUX
```
Done

seelabs replied:
That's great, thanks. One minor change I'd still like: instead of one function with the preprossor directives inside the function, can we change it so there are two separate functions and the preprossor directives control which function is compiled in?

undertome replied:
I missed this comment. What's the benefit vs having the directives inside the function?

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1350,19 +1500,106 @@ DatabaseShardImp::getCache(std::uint32_t seq)
+            // shares that byte count.
+            std::unordered_map<std::uint64_t, std::vector<std::string>>
+                uniqueCapacities(historicalPaths_.size());
+#endif
```
Done

undertome commented on src/ripple/net/ShardDownloader.md:
```diff
@@ -177,7 +177,7 @@ The `download_path` field of the `shard_db` entry is used to determine where to
 path=/var/lib/rippled/db/shards/nudb
 download_path=/var/lib/rippled/db/shards/
-max_size_gb=50
+max_historical_shards=50
```
Yes, the `max_size_gb` key is being removed. The decision came from @miguelportilla. This is a preexisting convention BTW, since the node store limits are determined by ledger counts and not drive capacity.

seelabs replied:
I defer to you and @miguelportilla on this. It's probably OK to break backward compatibility as few people are using shards right now.

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1402,23 +1573,363 @@ DatabaseShardImp::storeLedgerInShard(
+}
+
+void
+DatabaseShardImp::relocateOutdatedShards(std::lock_guard<std::mutex> const& lock)
```
clang format will reformat this line

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -292,10 +335,19 @@ DatabaseShardImp::prepareShard(std::uint32_t shardIndex)
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    bool const isHistoricalShard = shardIndex < shardBoundaryIndex(lock);
+    auto const numHistoricalShards = this->numHistoricalShards(lock);
```
Now that we've renamed the function, we should probably rename this variables. Maybe `numHistShards`.

undertome replied:
Done

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -688,23 +755,21 @@ DatabaseShardImp::import(Database& source)
             }
         }
 
+        auto numHistoricalShards = this->numHistoricalShards(lock);
```
Consider renaming the var.

undertome replied:
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1402,23 +1573,362 @@ DatabaseShardImp::storeLedgerInShard(
+
+        if (!filesystemIDs.insert(buffer.f_fsid).second)
+        {
+            JLOG(j_.error())
+                << "each 'historical_shard_path' must correspond to a "
+                   "distinct device or filesystem";
```
I think it would be worthwhile to output the duplicate path here, too. But now that I look at it, it would be even better to output both paths that are on the same device so the user can pick one. I think the best way to implement this would be to use a `map` similar to how you checked below for non-linux filesystems.

undertome replied:
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1402,23 +1573,362 @@ DatabaseShardImp::storeLedgerInShard(
+
+    for (auto const& path : historicalPaths_)
+    {
+        auto const availableSpace = boost::filesystem::space(path).available;
+
+        if (uniqueCapacities.find(availableSpace) != uniqueCapacities.end())
+            uniqueCapacities[availableSpace].push_back(path.string());
+        else
+            uniqueCapacities[availableSpace] = {path.string()};
```
You can do this in one line because the `[]` operator creates a default value (i.e. an empty vector) as soon as you invoke it.
```
uniqueCapacities[boost::filesystem::space(path).available].push_back(path.string());
```

undertome replied:
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.h:
```diff
@@ -242,6 +250,16 @@ class DatabaseShardImp : public DatabaseShard
+    // pending shards will be stored on the filesystem. A value of
+    // boost::none indicates that the corresponding shard is not held
+    // by the database.
+    boost::optional<std::uint32_t> latestShardIndex_;
+    boost::optional<std::uint32_t> secondLatestShardIndex_;
```
Instead of `latest` and `secondLatest`, what do you think about `last` and `penultimate`? I think it's fine and clear the way it is, but "penultimate" means "next to last", and it's a little shorter, sooooooooo......

undertome replied:
Think I'll opt to keep the names as they are.

ximinez replied:
> Think I'll opt to keep the names as they are.

I'm fine with that.

undertome commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1402,23 +1573,362 @@ DatabaseShardImp::storeLedgerInShard(
+        struct statvfs buffer;
+        if (statvfs(path.c_str(), &buffer))
+        {
+            JLOG(j_.error())
+                << "failed to acquire stats for a 'historical_shard_path'";
```
Done

ximinez commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1864,25 +1864,39 @@ DatabaseShardImp::checkHistoricalPaths() const
+                << "The following paths correspond to the same filesystem: "
+                << boost::algorithm::join(entry.second, ", ")
+                << ". Each configured historical storage path should"
+                   " be on a unique device or filesystem.";
```
Need to `return false;`

undertome replied:
Done

undertome commented:
> Looks like there's an issue with the VS build:
> 
> ```
> (WarnCompileDuplicatedFilename target) -> 
> 
>          C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.CppBuild.targets(944,5): warning MSB8027: Two or more files with the name of Manifest.cpp will produce outputs to the same location. This can lead to an incorrect build result.  The files involved are C:\Users\travis\build\ripple\rippled\src\ripple\app\misc\impl\Manifest.cpp, C:\Users\travis\build\ripple\rippled\src\ripple\rpc\handlers\Manifest.cpp. [C:\Users\travis\build\ripple\rippled\build.ms\rippled.vcxproj]
> ```
> 
> We've run into similar issues in the past and have solved it by renaming one of the files.

Let's address this in a separate PR since this one didn't change those files.

ximinez commented:
@seelabs I think that error is a non-fatal warning, and the actual CI failure was due to Travis being Travis. When I rebuilt that step, it succeeded. https://travis-ci.com/github/ripple/rippled/jobs/375010094
